File: gtci/02-fast-and-slow-pointers/python3/4-happy-number.py
class Solution:
    def find(self, num):
        slow = num
        fast = num
        while True:
            slow = self.calculate_square_sum(slow)
            fast = self.calculate_square_sum(self.calculate_square_sum(fast))

            if slow == fast:
                break

        return slow == 1

    def calculate_square_sum(self, num):
        square_sum = 0
        while num > 0:
            digit = num % 10
            square_sum += digit**2

            num //= 10

        return square_sum


if __name__ == "__main__":
    sol = Solution()
    assert sol.find(23) == True
    assert sol.find(12) == False

File: gtci/02-fast-and-slow-pointers/cpp/4-happy-number.cpp
#include <cassert>
#include <iostream>

using namespace std;

class Solution {
   public:
    bool find(int num) {
        int slow = num;
        int fast = num;

        do {
            slow = Solution::calculateSquareSum(slow);
            fast = Solution::calculateSquareSum(
                Solution::calculateSquareSum(fast));
        } while (slow != fast);

        return slow == 1;
    }

   private:
    static int calculateSquareSum(int num) {
        int squareSum = 0;

        while (num > 0) {
            const int digit = num % 10;
            squareSum += digit * digit;

            num /= 10;
        }

        return squareSum;
    }
};

int main() {
    Solution sol;
    assert(sol.find(23) == true);
    assert(sol.find(12) == false);

    cout << "All test cases passed." << endl;

    return 0;
}
